### PR TITLE
Fix daemon key file location

### LINKED ISSUE #10233: .docker/key.json: permission denied
On my dev environment, I start the docker daemon using `sudo`

In my `~/.docker` the trust file `key.json` is created when the daemon starts, owned by root.

Since a few days, I believe the client is also generating such a key and use it to validate pulls and pushes.

The issue is, as I'm using `sudo`, the client is trying to read the `key.json` file owned by root instead of created it's own.

So I can't even do a `docker ps`, I get 

```
FATA[0000] Error loading key file: unable to read key file /home/vagrant/.docker/key.json: open /home/vagrant/.docker/key.json: permission denied
```

### PR DESCRIPTION
**Pull request by dmcgowan**
Used placeholder instead of flag, can add the flag just need to bring over the documentation as well.

Fixes #10233

### CODE REVIEW

**Comment by icecrime** [comment]
Thanks for the quick patch! So if I understand correctly, client and daemon will now use two separate `key.json` files.

@vieux Can you tell us if this works for you?

File: integration-cli/docker_cli_daemon_test.go [hunk 1]
```diff
@@ -349,4 +351,24 @@ func TestDaemonVolumesBindsRefs(t *testing.T) {
 	}
 
 	logDone("daemon - bind refs in data-containers survive daemon restart")
+
```
Comment (tiborvass) [inline integration-cli/docker_cli_daemon_test.go]: @dmcgowan there's a `}` missing here

File: integration-cli/docker_cli_daemon_test.go [hunk 2]
```diff
@@ -349,4 +351,24 @@ func TestDaemonVolumesBindsRefs(t *testing.T) {
...
+func TestDaemonKeyGeneration(t *testing.T) {
+	os.Remove("/etc/docker/key.json")
```
Comment (thaJeztah) [inline integration-cli/docker_cli_daemon_test.go]: Probably silly question, but should these paths use getDaemonConfDir(). Probably moot, 'cause there's no Windows daemon (yet)

**Comment by tianon** [comment]
Had my machine not start docker on boot thanks to this last night. :cry:

**Comment by vieux** [comment]
Ok, it's working, 3 comments:

- The daemon key was moved to `/etc/docker`, it means, it wasn't working at first because I still had the `key.json` owned by root in my home, I had to remove it **manually**.

- On every system (even if you don't use `sudo`) the key is going to change because we moved from `/root/.docker/key.json` to `/etc/docker` I hope it won't impact too much swarm as the key is the daemon ID.

- Why not `/var/lib/docker` ?

**Comment by jfrazelle** [comment]
I nuke my graph dir all the time, so maybe not /var/lib/docker

**Comment by tiborvass** [comment]
I think it's fine to remove the key manually this time, since it was not released with the bug.

**Comment by dmcgowan** [comment]
@vieux the key changing does concern me, what will be the impact to Swarm if the daemon identifier changes?  If this identifier should not change, the file will need to be moved instead of removed and regenerated.

As for `/var/lib/docker`, I agree with @jfrazelle and there is other discussion here https://github.com/docker/docker/pull/8265#issuecomment-68383827

**Comment by icecrime** [comment]
After discussing with @dmcgowan and @vieux:

* It's *not* ok to have to remove the key manually, because 1.4 daemon was already creating it (but client wasn't using it though)
* Preserving key stability (i.e., that existing daemons don't generate a new one when upgrading to 1.5) would be appreciated, although not strictly necessary

This seems to advocate for the following:

* The client will generate its own key in a *different* file (`key-client.json` or similar): this fixes the client unable to read the key because of permissions issue
* The daemon will detect a previously existing `key.json` in the home directory and move it to its new storage path (`/etc/docker/`): this fixes key stability

**Comment by vieux** [comment]
@tiborvass yes it was, the key was created in the `$HOME` owned by root since 1.4

My suggestions would be to
- use `~/.docker/key-client.json` or `~/.docker/key-cli.json` in the client, so it fixes (1)
- At the daemon start, we check  `~/.docker/key.json`, because of docker 1.4, this file has to be the daemon one, so if there is such a file, we move it to `/etc/docker/key.json` or `/etc/docker/key-daemon.json`, it fixes (2)

**Comment by crosbymichael** [comment]
I guess moving sounds reasonable to me since this is in a `.docker` dir

**Comment by tiborvass** [comment]
@vieux oh my bad! Yeah in that case we need to fix this. Moving is okay for me.

**Comment by tianon** [comment]
So, what about the case where on system bootup, `/etc/init.d/docker` has no `HOME` so the key goes to `/.docker/key.json`, but when you do `sudo /etc/init.d/docker restart` it now has `HOME` (`/root`, thanks to `sudo`) so the key goes to `/root/.docker/key.json` instead?

**Comment by crosbymichael** [comment]
ref #9468

**Comment by dmcgowan** [comment]
The daemon will use /etc/docker, only the client will rely on the HOME environment variable. This should prevent any accidental creation of "/.docker" from init scripts.  Is there some other parameters of that discussion I am overlooking?

**Comment by tianon** [comment]
Just the fact that 1.4.1 already creates "/.docker" today, and so it should
be part of the "move the existing file" logic IMO.

**Comment by dmcgowan** [comment]
Should have an update for this within the hour.  The migration logic is as follows...
If /etc/docker/key.json does not exist and $HOME/.docker/key.json exists and is owned by the daemon user, then create /etc/docker/key.json, copy contents from $HOME/.docker/key.json, and remove $HOME/.docker/key.json.

This should cover the case of $HOME == "" and ensure that /etc/docker/key.json is not overwritten once created.  I also think I can just scrap separating "key.json" and "key-cli.json".  If the client gets run as root and then not as root, we will still have the permission problem though.  I am still not quite sure what the best way to correct that behavior is other than ensuring we delay creation of the key until a command which needs it is run.  This is relevant to some installation instructions.

**Comment by thaJeztah** [comment]
Perhaps check both `$HOME/.docker/key.json` *and* `/.docker/key.json` (in that order), in case `$HOME` wasn't set previously, but now is (is that case possible?); If I understand correctly, this is just a one-time migration step, so searching a few more locations wouldn't be costly.

**Comment by tianon** [comment]
I agree with Sebastiaan here, although I'd recommend we swap the order.
It's very easy to get your Daemon starting up with HOME set to something,
but it's difficult to get it to start up without it (since fresh boot will
have it not set, but then any restart of the daemon will set it
inadvertently).

**Comment by icecrime** [comment]
> *"... and is owned by the daemon user..."*

I'm not sure we need this condition: as far as I can tell, there is no way that this file was created by anything but the daemon.

**Comment by dmcgowan** [comment]
@thaJeztah In that case I would prefer that administrators clean it up manually.  What was the final solution to the issue creating /.docker?  Was it set $HOME and migrate the key or just set $HOME and remove/ignore previous key value?

**Comment by dmcgowan** [comment]
@icecrime I would prefer to be safer here, it is possible that it was created by 1.5 client, then later running a 1.5 daemon could migrate the client's key.

**Comment by tianon** [comment]
There wasn't a solution yet to the daemon creating "/.docker"; that's
exactly what we're discussing.  All my machines have this directory still,
whether they're running a release or master.

**Comment by thaJeztah** [comment]
Problem is of course, which one is the "right" key.json that should be copied; Ye oldest or most recent?

**Comment by dmcgowan** [comment]
@thaJeztah as far as I know swarm is the only use case that is relying on the identity of the key.  Picking the wrong key would have the effect of changing what the output from `docker info` is and the identity for swarm.  Not to say we shouldn't try to preserve it, just not sure of what current scenarios it would be noticed.

**Comment by thaJeztah** [comment]
@dmcgowan right. Perhaps if *both* locations are found, output a message with instructions, something like;

    ***warning***
    Unable to automatically migrate existing key. 
    Existing keys were found at the following locations;

    /.docker/key.json
    /foo/key.json

    To use your existing key with docker, you have to move the key to /etc/docker/.....
    and restart the daemon (?).

**Comment by crosbymichael** [comment]
I'm not worried about the swarm usecase and jumping though hoops to support it.  It's not released yet so i don't see a problem right now.

The thing I really care about is not leaving behind a much of trash on the filesystem with various owners and permissions .

**Comment by vieux** [comment]
It might not be only swarm, `docker info` displays the ID since docker 1.4. It's in the API as well, other tools can depend on it.

**Comment by dmcgowan** [comment]
@vieux it is a valid point and I think migrating should sufficiently support existing keys. I am comfortable only migrating a key at $HOME/.docker/key.json since that is the key that previous version of the daemon would expect to use.  Any other key files that got created are up to the user to clean up, we will not start to manage them since we were not before.

PTAL at the migration key that is the logic we discussed and can agree on.

File: docker/flags.go [hunk 3]
```diff
@@ -28,6 +28,13 @@ func getHomeDir() string {
 	return os.Getenv("HOME")
 }
 
+func getDaemonConfDir() string {
+	if runtime.GOOS == "windows" {
+		return filepath.Join(os.Getenv("USERPROFILE"), ".docker")
```
Comment (vieux) [inline docker/flags.go]: we should remember to change this when the daemon will run on windows, otherwise we will get the same issue.

Comment (dmcgowan) [inline docker/flags.go]: Is there a standard TODO marker for future windows updates?  I added this mostly as a placeholder and would expect the Windows developers to know what is best.

Comment (vieux) [inline docker/flags.go]: We juste use `// TODO:`

**Comment by vieux** [comment]
We could add a test for the migration, but beside that LGTM

File: docker/flags.go (hunk 3 above)
Comment (tiborvass) [inline docker/flags.go]: There's also `FIXME` sometimes (we're not that consistent)

File: docker/daemon.go [hunk 4]
```diff
@@ -28,6 +34,38 @@ func init() {
 	registryCfg.InstallFlags()
 }
 
+func migrateKey() error {
+	// Migrate trust key if exists at ~/.docker/key.json and owned by current user
+	oldPath := filepath.Join(getHomeDir(), ".docker", defaultTrustKeyFile)
+	newPath := filepath.Join(getDaemonConfDir(), defaultTrustKeyFile)
+	if _, err := os.Stat(newPath); os.IsNotExist(err) && utils.IsFileOwner(oldPath) {
+		if err := os.MkdirAll(getDaemonConfDir(), os.FileMode(0644)); err != nil {
+			return fmt.Errorf("Unable to create daemon configuraiton directory: %s", err)
```
Comment (icecrime) [inline docker/daemon.go]: s/configuraiton/configuration/

File: docker/daemon.go [hunk 5]
```diff
@@ -28,6 +34,38 @@ func init() {
...
+		oldFile, err := os.Open(oldPath)
+		if err != nil {
+			return fmt.Errorf("error opening open key file %q: %s", oldPath, err)
+		}
+
+		if _, err := io.Copy(newFile, oldFile); err != nil {
+			return fmt.Errorf("error copying key: %s", err)
+		}
+
+		oldFile.Close()
```
Comment (icecrime) [inline docker/daemon.go]: Small nitpick, but maybe we should `defer` the call to the `Close()` just before the `io.Copy`.

File: docker/daemon.go [hunk 6]
```diff
@@ -28,6 +34,38 @@ func init() {
...
+		log.Debugf("Migrated key from %s to %s", oldPath, newPath)
```
Comment (icecrime) [inline docker/daemon.go]: It should only happen once, I think it's worth having a `log.Infof` for this.

File: docker/daemon.go (hunk 5 above)
Comment (dmcgowan) [inline docker/daemon.go]: Reason I didn't defer the close is that I am calling os.Remove(oldPath), I could defer that as well

File: docker/daemon.go [hunk 7]
```diff
@@ -28,6 +34,38 @@ func init() {
...
+
+		newFile, err := os.OpenFile(newPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
+		if err != nil {
+			return fmt.Errorf("error creating key file %q: %s", newPath, err)
+		}
+		defer newFile.Close()
+
+		oldFile, err := os.Open(oldPath)
+		if err != nil {
+			return fmt.Errorf("error opening open key file %q: %s", oldPath, err)
```
Comment (tiborvass) [inline docker/daemon.go]: opening ~~open~~

**Comment by icecrime** [comment]
Small nitpicks aside, tested and LGTM!

File: docker/daemon.go (hunk 5 above)
Comment (crosbymichael) [inline docker/daemon.go]: well you leak the file on a `io.Copy` error

**Comment by tiborvass** [comment]
LGTM

**Comment by icecrime** [comment]
LGTM

**Pull request merged**